db/db_impl.h

```diff
@@ -510,9 +510,11 @@ class DBImpl : public DB {
     uint64_t log_number_;
     std::string name_;
     WriteBatch* batch_;
+    // The seq number of the first key in the batch
+    SequenceNumber seq_;
     explicit RecoveredTransaction(const uint64_t log, const std::string& name,
-                                  WriteBatch* batch)
-        : log_number_(log), name_(name), batch_(batch) {}
+                                  WriteBatch* batch, SequenceNumber seq)
+        : log_number_(log), name_(name), batch_(batch), seq_(seq) {}
```

Comment:
For backward compatibility we would have to keep that logic for a long time. Also MarkLogAsContainingPrepSection is still required because during recovery we need to recover (at least) the list of prepared but not committed transaction. If the log that has such transaction is gone then recovery logic would fail.
A better design however would be to use a separate column family for txn metadata and avoid any kind of marker in the WAL. We however decided to stick with the existing marker-based approach to be able to ship the feature faster.

Reply:
Make sense to keep them for backward compatibility. 

Regarding MarkLogAsContainingPrepSection, now that prepare write to memtable, the log will be kept if memtable has not been flush, or we can throw away the log if memtable is flushed. In the later case, what's missing from the flushed memtable is the prepare marker to restore prepared transaction state. I'm wondering how this is handle in this PR.

Reply:
For the last question, I mean how do you handle the case if 
* txn1 write to memtable1
* memtable1 get flushed.
* recovery go through the log, sees txn1 but since memtable1 has flushed, it skips txn1.

Reply:
currently there are two mechanisms to keep a grip on old logs:
1. keep the old logs that have prepared data of what it is committed to the memtable
2. keep the old logs that have prepared but not committed data
The former is not necessary for WritePrepared since the log that backs up the memtable is the same as the current log.
The latter is reused to be able to restore the list of prepared txns. It is done in Prepare() and Commit() methods in PessimisticTxn where after the prepare and commit the txn tell the DBImpl whether it needs to keep an old log.

Reply:
OK. Just want to double check flush will be fine and it is.